internal/api/creator.go

```diff
@@ -0,0 +1,200 @@
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+
+	kmmv1beta1 "github.com/kubernetes-sigs/kernel-module-management/api/v1beta1"
+	buildutils "github.com/kubernetes-sigs/kernel-module-management/internal/build/utils"
+	"github.com/kubernetes-sigs/kernel-module-management/internal/utils"
+)
+
+//go:generate mockgen -source=creator.go -package=api -destination=mock_creator.go ModuleLoaderDataCreator,moduleLoaderDataCreatorHelper
+
+type ModuleLoaderDataCreator interface {
+	FromModule(mod *kmmv1beta1.Module, kernelVersion string) (*ModuleLoaderData, error)
+}
+
+type moduleLoaderDataCreator struct {
+	helper moduleLoaderDataCreatorHelper
+}
+
+func NewModuleLoaderDataCreator() ModuleLoaderDataCreator {
+	return &moduleLoaderDataCreator{
+		helper: &moduleLoaderDataCreatorHelperImpl{
+			sectionGetter: &sectionGetterImpl{},
+		},
+	}
+}
+
+func (f *moduleLoaderDataCreator) FromModule(mod *kmmv1beta1.Module, kernelVersion string) (*ModuleLoaderData, error) {
+	mappings := mod.Spec.ModuleLoader.Container.KernelMappings
+	foundMapping, err := f.helper.findKernelMapping(mappings, kernelVersion)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find mapping for kernel %s: %v", kernelVersion, err)
+	}
+	mld, err := f.helper.prepareModuleLoaderData(foundMapping, mod, kernelVersion)
+	if err != nil {
+		return nil, fmt.Errorf("failed to prepare module loader data for kernel %s: %v", kernelVersion, err)
+	}
+
+	err = f.helper.replaceTemplates(mld)
+	if err != nil {
+		return nil, fmt.Errorf("failed to replace templates in module loader data for kernel %s: %v", kernelVersion, err)
+	}
+	return mld, nil
+}
+
+type moduleLoaderDataCreatorHelper interface {
+	findKernelMapping(mappings []kmmv1beta1.KernelMapping, kernelVersion string) (*kmmv1beta1.KernelMapping, error)
+	prepareModuleLoaderData(mapping *kmmv1beta1.KernelMapping, mod *kmmv1beta1.Module, kernelVersion string) (*ModuleLoaderData, error)
+	replaceTemplates(mld *ModuleLoaderData) error
+}
+
+type moduleLoaderDataCreatorHelperImpl struct {
+	sectionGetter sectionGetter
+}
+
+func (kh *moduleLoaderDataCreatorHelperImpl) findKernelMapping(mappings []kmmv1beta1.KernelMapping, kernelVersion string) (*kmmv1beta1.KernelMapping, error) {
+	for _, m := range mappings {
+		if m.Literal != "" && m.Literal == kernelVersion {
+			return &m, nil
+		}
+
+		if m.Regexp == "" {
+			continue
+		}
+
+		if matches, err := regexp.MatchString(m.Regexp, kernelVersion); err != nil {
+			return nil, fmt.Errorf("could not match regexp %q against kernel %q: %v", m.Regexp, kernelVersion, err)
+		} else if matches {
+			return &m, nil
+		}
+	}
+
+	return nil, errors.New("no suitable mapping found")
+}
+
+func (kh *moduleLoaderDataCreatorHelperImpl) prepareModuleLoaderData(mapping *kmmv1beta1.KernelMapping, mod *kmmv1beta1.Module, kernelVersion string) (*ModuleLoaderData, error) {
+	var err error
+
+	mld := &ModuleLoaderData{}
+	// prepare the build
+	if mapping.Build != nil || mod.Spec.ModuleLoader.Container.Build != nil {
```

Comment:
what do you think about moving the checks for kenrelMapping sign/buid and module sign/build into the getRelevantBuidl/Sign functions? This will make prepareModuleLoader code cleaner, and those functions are returning pointer anyway